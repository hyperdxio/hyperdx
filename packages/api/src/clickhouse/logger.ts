import {
    Logger as _CHLogger,
  } from '@clickhouse/client';
  import {
    LogParams as _CHLogParams,
    ErrorLogParams as _CHErrorLogParams,
  } from '@clickhouse/client/dist/logger';
  import logger from '@/utils/logger';

export class CHLogger implements _CHLogger {
    debug({ module, message, args }: _CHLogParams): void {
      logger.debug({
        type: '@clickhouse/client',
        module,
        message,
        ...args,
      });
    }
  
    trace({ module, message, args }: _CHLogParams) {
      // TODO: trace level ??
      logger.info({
        type: '@clickhouse/client',
        module,
        message,
        ...args,
      });
    }
  
    info({ module, message, args }: _CHLogParams): void {
      logger.info({
        type: '@clickhouse/client',
        module,
        message,
        ...args,
      });
    }
  
    warn({ module, message, args }: _CHLogParams): void {
      logger.warn({
        type: '@clickhouse/client',
        module,
        message,
        ...args,
      });
    }
  
    error({ module, message, args, err }: _CHErrorLogParams): void {
      logger.error({
        type: '@clickhouse/client',
        module,
        message,
        ...args,
        err,
      });
    }
  }