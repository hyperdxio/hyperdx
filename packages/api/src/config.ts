const env = process.env;
export const NODE_ENV = env.NODE_ENV as string;

export const AGGREGATOR_API_URL = env.AGGREGATOR_API_URL as string;
export const APP_TYPE = env.APP_TYPE as 'api' | 'aggregator' | 'scheduled-task';
export const CLICKHOUSE_HOST = env.CLICKHOUSE_HOST as string;
export const CLICKHOUSE_PASSWORD = env.CLICKHOUSE_PASSWORD as string;
export const CLICKHOUSE_USER = env.CLICKHOUSE_USER as string;
export const CODE_VERSION = env.CODE_VERSION as string;
export const COOKIE_DOMAIN = env.COOKIE_DOMAIN as string; // prod ONLY
export const EXPRESS_SESSION_SECRET = env.EXPRESS_SESSION_SECRET as string;
export const FRONTEND_URL = env.FRONTEND_URL as string;
export const HYPERDX_API_KEY = env.HYPERDX_API_KEY as string;
export const HYPERDX_LOG_LEVEL = env.HYPERDX_LOG_LEVEL as string;
export const INGESTOR_API_URL = env.INGESTOR_API_URL as string;
export const IS_CI = NODE_ENV === 'ci';
export const IS_DEV = NODE_ENV === 'development';
export const IS_PROD = NODE_ENV === 'production';
export const MINER_API_URL = env.MINER_API_URL as string;
export const MONGO_URI = env.MONGO_URI as string;
export const OTEL_EXPORTER_OTLP_ENDPOINT =
  env.OTEL_EXPORTER_OTLP_ENDPOINT as string;
export const OTEL_SERVICE_NAME = env.OTEL_SERVICE_NAME as string;
export const PORT = Number.parseInt(env.PORT as string);
export const REDIS_URL = env.REDIS_URL as string;
export const SERVER_URL = env.SERVER_URL as string;
export const USAGE_STATS_ENABLED = env.USAGE_STATS_ENABLED !== 'false';
