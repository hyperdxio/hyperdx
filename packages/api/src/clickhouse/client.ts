import {
    Logger as _CHLogger,
    createClient,
  } from '@clickhouse/client';
import { CHLogger } from './logger'
import * as config from '@/config';
import ms from 'ms';

export const client = createClient({
    host: config.CLICKHOUSE_HOST,
    username: config.CLICKHOUSE_USER,
    password: config.CLICKHOUSE_PASSWORD,
    request_timeout: ms('1m'),
    compression: {
      request: false,
      response: false, // has to be off to enable streaming
    },
    keep_alive: {
      enabled: true,
      // should be slightly less than the `keep_alive_timeout` setting in server's `config.xml`
      // default is 3s there, so 2500 milliseconds seems to be a safe client value in this scenario
      // another example: if your configuration has `keep_alive_timeout` set to 60s, you could put 59_000 here
      socket_ttl: 60000,
      retry_on_expired_socket: true,
    },
    clickhouse_settings: {
      connect_timeout: ms('1m') / 1000,
      date_time_output_format: 'iso',
      max_download_buffer_size: (10 * 1024 * 1024).toString(), // default
      max_download_threads: 32,
      max_execution_time: ms('2m') / 1000,
    },
    log: {
      LoggerClass: CHLogger,
    },
  });
  