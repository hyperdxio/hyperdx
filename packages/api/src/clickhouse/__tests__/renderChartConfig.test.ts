// TODO: we might want to move this test file to common-utils package

import { ChSql, ClickhouseClient } from '@hyperdx/common-utils/dist/clickhouse';
import { getMetadata } from '@hyperdx/common-utils/dist/metadata';
import { renderChartConfig } from '@hyperdx/common-utils/dist/renderChartConfig';
import _ from 'lodash';
import ms from 'ms';

import { MetricsDataType } from '@/../../common-utils/dist/types';
import * as config from '@/config';
import { createTeam } from '@/controllers/team';
import {
  bulkInsertLogs,
  bulkInsertMetricsGauge,
  bulkInsertMetricsSum,
  DEFAULT_DATABASE,
  DEFAULT_LOGS_TABLE,
  DEFAULT_METRICS_TABLE,
  getServer,
} from '@/fixtures';
import Connection from '@/models/connection';
import { Source } from '@/models/source';

describe('renderChartConfig', () => {
  const server = getServer();

  const now = new Date('2022-01-05').getTime();
  let team, connection, logSource, metricSource, metadata;
  let clickhouseClient: ClickhouseClient;

  const queryData = async (chsql: ChSql) => {
    try {
      const res = await clickhouseClient.query<'JSON'>({
        query: chsql.sql,
        query_params: chsql.params,
        format: 'JSON',
      });
      const json = await res.json();
      return json.data;
    } catch (err) {
      console.error('[ClickhouseClient] Error:', err);
      throw err;
    }
  };

  beforeAll(async () => {
    await server.start();
  });

  afterAll(async () => {
    await server.stop();
  });

  beforeEach(async () => {
    team = await createTeam({ name: 'My Team' });
    connection = await Connection.create({
      team: team._id,
      name: 'Default',
      host: config.CLICKHOUSE_HOST,
      username: config.CLICKHOUSE_USER,
      password: config.CLICKHOUSE_PASSWORD,
    });
    logSource = await Source.create({
      kind: 'log',
      team: team._id,
      from: {
        databaseName: DEFAULT_DATABASE,
        tableName: DEFAULT_LOGS_TABLE,
      },
      timestampValueExpression: 'Timestamp',
      connection: connection.id,
      name: 'Logs',
    });
    metricSource = await Source.create({
      kind: 'metric',
      team: team._id,
      from: {
        databaseName: DEFAULT_DATABASE,
        tableName: '',
      },
      timestampValueExpression: 'TimeUnix',
      connection: connection.id,
      name: 'OTLPMetrics',
    });
    clickhouseClient = new ClickhouseClient({
      host: connection.host,
      username: connection.username,
      password: connection.password,
    });
    metadata = getMetadata(clickhouseClient);
  });

  afterEach(async () => {
    await server.clearDBs();
    jest.clearAllMocks();
  });

  describe('Query Events', () => {
    it('simple select + where query logs', async () => {
      const now = new Date('2023-11-16T22:12:00.000Z');
      await bulkInsertLogs([
        {
          ServiceName: 'api',
          Timestamp: now,
          SeverityText: 'error',
          Body: 'Oh no! Something went wrong!',
        },
        {
          ServiceName: 'api',
          Timestamp: now,
          SeverityText: 'info',
          Body: 'This is a test message.',
        },
      ]);
      const query = await renderChartConfig(
        {
          select: [
            {
              valueExpression: 'Body',
            },
          ],
          from: logSource.from,
          where: `SeverityText = 'error'`,
          timestampValueExpression: 'Timestamp',
          connection: connection.id,
        },
        metadata,
      );

      const resp = await clickhouseClient
        .query<'JSON'>({
          query: query.sql,
          query_params: query.params,
          format: 'JSON',
        })
        .then(res => res.json<any>());
      expect(resp.data).toMatchSnapshot();
    });

    it('simple select + group by query logs', async () => {
      const now = new Date('2023-11-16T22:12:00.000Z');
      await bulkInsertLogs([
        {
          ServiceName: 'api',
          Timestamp: now,
          SeverityText: 'error',
          Body: 'Oh no! Something went wrong!',
        },
        {
          ServiceName: 'app',
          Timestamp: now,
          SeverityText: 'info',
          Body: 'This is a test message.',
        },
      ]);
      const query = await renderChartConfig(
        {
          select: [
            {
              valueExpression: 'ServiceName',
            },
            {
              valueExpression: 'count()',
              alias: 'count',
            },
          ],
          from: logSource.from,
          where: '',
          timestampValueExpression: 'Timestamp',
          connection: connection.id,
          groupBy: 'ServiceName',
        },
        metadata,
      );
      expect(await queryData(query)).toMatchSnapshot();
    });

    // TODO: add more tests (including events chart, using filters, etc)
  });

  describe('Query Metrics', () => {
    beforeEach(async () => {
      const gaugePointsA = [
        { value: 50, timestamp: now },
        { value: 25, timestamp: now + ms('1m') },
        { value: 12.5, timestamp: now + ms('2m') },
        { value: 6.25, timestamp: now + ms('3m') },
        { value: 100, timestamp: now + ms('6m') },
        { value: 75, timestamp: now + ms('7m') },
        { value: 10, timestamp: now + ms('8m') },
        { value: 80, timestamp: now + ms('9m') },
      ].map(point => ({
        MetricName: 'test.cpu',
        ResourceAttributes: {
          host: 'host1',
          ip: '127.0.0.1',
        },
        Value: point.value,
        TimeUnix: new Date(point.timestamp),
      }));
      const gaugePointsB = [
        { value: 1, timestamp: now },
        { value: 2, timestamp: now + ms('1m') },
        { value: 3, timestamp: now + ms('2m') },
        { value: 4, timestamp: now + ms('3m') },
        { value: 5, timestamp: now + ms('6m') },
        { value: 6, timestamp: now + ms('7m') },
        { value: 5, timestamp: now + ms('8m') },
        { value: 4, timestamp: now + ms('9m') },
      ].map(point => ({
        MetricName: 'test.cpu',
        ResourceAttributes: {
          host: 'host2',
          ip: '127.0.2',
        },
        Value: point.value,
        TimeUnix: new Date(point.timestamp),
      }));
      // Rate: 8, 1, 8, 25
      const sumPointsA = [
        { value: 0, timestamp: now - ms('1m') }, // 0
        { value: 1, timestamp: now },
        { value: 8, timestamp: now + ms('4m') }, // 8
        { value: 8, timestamp: now + ms('6m') },
        { value: 9, timestamp: now + ms('9m') }, // 9
        { value: 15, timestamp: now + ms('11m') },
        { value: 17, timestamp: now + ms('14m') }, // 17
        { value: 32, timestamp: now + ms('16m') },
        { value: 42, timestamp: now + ms('19m') }, // 42
      ].map(point => ({
        MetricName: 'test.users',
        ResourceAttributes: {
          host: 'host1',
          ip: '127.0.0.1',
        },
        Value: point.value,
        TimeUnix: new Date(point.timestamp),
        IsMonotonic: true,
        AggregationTemporality: 2, // Cumulative
      }));
      // Rate: 11, 78, 5805, 78729
      // Sum: 12, 79, 5813, 78754
      const sumPointsB = [
        { value: 3, timestamp: now - ms('1m') }, // 3
        { value: 3, timestamp: now },
        { value: 14, timestamp: now + ms('4m') }, // 14
        { value: 15, timestamp: now + ms('6m') },
        { value: 92, timestamp: now + ms('9m') }, // 92
        { value: 653, timestamp: now + ms('11m') },
        { value: 5897, timestamp: now + ms('14m') }, // 5897
        { value: 9323, timestamp: now + ms('16m') },
        { value: 84626, timestamp: now + ms('19m') }, // 84626
      ].map(point => ({
        MetricName: 'test.users',
        ResourceAttributes: {
          host: 'host2',
          ip: '127.0.0.1',
        },
        Value: point.value,
        TimeUnix: new Date(point.timestamp),
        IsMonotonic: true,
        AggregationTemporality: 2, // Cumulative
      }));
      const sumPointsC = [
        { value: 0, timestamp: now - ms('1m') }, // 0
        { value: 1, timestamp: now },
        { value: 8, timestamp: now + ms('1m') }, // 8
        { value: 0, timestamp: now + ms('2m') },
        { value: 7, timestamp: now + ms('2m') },
        { value: 7, timestamp: now + ms('10m') }, // 9
        { value: 15, timestamp: now + ms('12m') },
        { value: 17, timestamp: now + ms('14m') }, // 17
        { value: 0, timestamp: now + ms('16m') },
        { value: 42, timestamp: now + ms('19m') }, // 42
      ].map(point => ({
        MetricName: 'counter.reset',
        ResourceAttributes: {
          host: 'host3',
          ip: '127.0.1',
        },
        Value: point.value,
        TimeUnix: new Date(point.timestamp),
        IsMonotonic: true,
        AggregationTemporality: 2, // Cumulative
      }));
      const sumPointsD = [
        { value: 0, timestamp: now - ms('1m') }, // 0
        { value: 1, timestamp: now },
        { value: 8, timestamp: now + ms('1m') }, // 8
        { value: 0, timestamp: now + ms('2m') },
        { value: 7, timestamp: now + ms('2m') },
        { value: 7, timestamp: now + ms('10m') }, // 9
        { value: 15, timestamp: now + ms('12m') },
        { value: 17, timestamp: now + ms('14m') }, // 17
        { value: 0, timestamp: now + ms('16m') },
        { value: 42, timestamp: now + ms('19m') }, // 42
      ].map(point => ({
        MetricName: 'counter.min_reset',
        ResourceAttributes: {
          host: 'MIN_VARIANT_0',
          ip: '127.0.1',
        },
        Value: point.value,
        TimeUnix: new Date(point.timestamp),
        IsMonotonic: true,
        AggregationTemporality: 2, // Cumulative
      }));
      const sumPointsE = [
        { value: 0, timestamp: now - ms('1m') },
        { value: 2, timestamp: now },
        { value: 9, timestamp: now + ms('1m') },
        { value: 0, timestamp: now + ms('2m') },
        { value: 15, timestamp: now + ms('2m') },
        { value: 25, timestamp: now + ms('10m') },
        { value: 35, timestamp: now + ms('12m') },
        { value: 57, timestamp: now + ms('14m') },
        { value: 0, timestamp: now + ms('16m') },
        { value: 92, timestamp: now + ms('19m') },
      ].map(point => ({
        MetricName: 'counter.min_reset',
        ResourceAttributes: {
          host: 'MAX_VARIANT_1',
          ip: '127.0.1',
        },
        Value: point.value,
        TimeUnix: new Date(point.timestamp),
        IsMonotonic: true,
        AggregationTemporality: 2, // Cumulative
      }));

      await Promise.all([
        bulkInsertMetricsGauge([...gaugePointsA, ...gaugePointsB]),
        bulkInsertMetricsSum([
          ...sumPointsA,
          ...sumPointsB,
          ...sumPointsC,
          ...sumPointsD,
          ...sumPointsE,
        ]),
      ]);
    });

    it('single max/avg/sum gauge', async () => {
      const avgQuery = await renderChartConfig(
        {
          select: [
            {
              aggFn: 'avg',
              metricName: 'test.cpu',
              metricType: MetricsDataType.Gauge,
              valueExpression: 'Value',
            },
          ],
          from: metricSource.from,
          where: '',
          metricTables: {
            sum: DEFAULT_METRICS_TABLE.SUM,
            gauge: DEFAULT_METRICS_TABLE.GAUGE,
            histogram: DEFAULT_METRICS_TABLE.HISTOGRAM,
          },
          dateRange: [new Date(now), new Date(now + ms('10m'))],
          granularity: '5 minute',
          timestampValueExpression: metricSource.timestampValueExpression,
          connection: connection.id,
        },
        metadata,
      );
      expect(await queryData(avgQuery)).toMatchSnapshot();
      const maxQuery = await renderChartConfig(
        {
          select: [
            {
              aggFn: 'max',
              metricName: 'test.cpu',
              metricType: MetricsDataType.Gauge,
              valueExpression: 'Value',
            },
          ],
          from: metricSource.from,
          where: '',
          metricTables: {
            sum: DEFAULT_METRICS_TABLE.SUM,
            gauge: DEFAULT_METRICS_TABLE.GAUGE,
            histogram: DEFAULT_METRICS_TABLE.HISTOGRAM,
          },
          dateRange: [new Date(now), new Date(now + ms('10m'))],
          granularity: '5 minute',
          timestampValueExpression: metricSource.timestampValueExpression,
          connection: connection.id,
        },
        metadata,
      );
      expect(await queryData(maxQuery)).toMatchSnapshot();
      const sumQuery = await renderChartConfig(
        {
          select: [
            {
              aggFn: 'sum',
              metricName: 'test.cpu',
              metricType: MetricsDataType.Gauge,
              valueExpression: 'Value',
            },
          ],
          from: metricSource.from,
          where: '',
          metricTables: {
            sum: DEFAULT_METRICS_TABLE.SUM,
            gauge: DEFAULT_METRICS_TABLE.GAUGE,
            histogram: DEFAULT_METRICS_TABLE.HISTOGRAM,
          },
          dateRange: [new Date(now), new Date(now + ms('10m'))],
          granularity: '5 minute',
          timestampValueExpression: metricSource.timestampValueExpression,
          connection: connection.id,
        },
        metadata,
      );
      expect(await queryData(sumQuery)).toMatchSnapshot();
    });

    it('single avg gauge with where', async () => {
      const query = await renderChartConfig(
        {
          select: [
            {
              aggFn: 'avg',
              metricName: 'test.cpu',
              metricType: MetricsDataType.Gauge,
              valueExpression: 'Value',
            },
          ],
          from: metricSource.from,
          where: `ResourceAttributes['host'] = 'host1'`,
          whereLanguage: 'sql',
          metricTables: {
            sum: DEFAULT_METRICS_TABLE.SUM,
            gauge: DEFAULT_METRICS_TABLE.GAUGE,
            histogram: DEFAULT_METRICS_TABLE.HISTOGRAM,
          },
          dateRange: [new Date(now), new Date(now + ms('10m'))],
          granularity: '5 minute',
          timestampValueExpression: metricSource.timestampValueExpression,
          connection: connection.id,
        },
        metadata,
      );
      expect(await queryData(query)).toMatchSnapshot();
    });

    it('single avg gauge with group-by', async () => {
      const query = await renderChartConfig(
        {
          select: [
            {
              aggFn: 'avg',
              metricName: 'test.cpu',
              metricType: MetricsDataType.Gauge,
              valueExpression: 'Value',
            },
          ],
          from: metricSource.from,
          where: '',
          metricTables: {
            sum: DEFAULT_METRICS_TABLE.SUM,
            gauge: DEFAULT_METRICS_TABLE.GAUGE,
            histogram: DEFAULT_METRICS_TABLE.HISTOGRAM,
          },
          dateRange: [new Date(now), new Date(now + ms('10m'))],
          granularity: '5 minute',
          groupBy: `ResourceAttributes['host']`,
          timestampValueExpression: metricSource.timestampValueExpression,
          connection: connection.id,
        },
        metadata,
      );
      expect(await queryData(query)).toMatchSnapshot();
    });

    it('single sum rate', async () => {
      const query = await renderChartConfig(
        {
          select: [
            {
              aggFn: 'sum',
              metricName: 'test.users',
              metricType: MetricsDataType.Sum,
              valueExpression: 'Value',
            },
          ],
          from: metricSource.from,
          where: '',
          metricTables: {
            sum: DEFAULT_METRICS_TABLE.SUM,
            gauge: DEFAULT_METRICS_TABLE.GAUGE,
            histogram: DEFAULT_METRICS_TABLE.HISTOGRAM,
          },
          dateRange: [new Date(now), new Date(now + ms('20m'))],
          granularity: '5 minute',
          timestampValueExpression: metricSource.timestampValueExpression,
          connection: connection.id,
        },
        metadata,
      );
      expect(await queryData(query)).toMatchSnapshot();
    });

    it('handles counter resets correctly for sum metrics', async () => {
      const query = await renderChartConfig(
        {
          select: [
            {
              aggFn: 'sum',
              metricName: 'counter.reset',
              metricType: MetricsDataType.Sum,
              valueExpression: 'Value',
            },
          ],
          from: metricSource.from,
          where: '',
          metricTables: {
            sum: DEFAULT_METRICS_TABLE.SUM,
            gauge: DEFAULT_METRICS_TABLE.GAUGE,
            histogram: DEFAULT_METRICS_TABLE.HISTOGRAM,
          },
          dateRange: [new Date(now), new Date(now + ms('20m'))],
          granularity: '10 minute',
          timestampValueExpression: metricSource.timestampValueExpression,
          connection: connection.id,
        },
        metadata,
      );
      expect(await queryData(query)).toMatchSnapshot();
    });

    // FIXME: here are the expected values
    // [0, 1, 8, 8, 15, 15, 23, 25, 25, 67]
    // [0, 2, 9, 9, 24, 34, 44, 66, 66, 158]
    // min -> [15, 52]
    // max -> [24, 134]
    it.skip('calculates min_rate/max_rate correctly for sum metrics', async () => {
      const minQuery = await renderChartConfig(
        {
          select: [
            {
              aggFn: 'min',
              metricName: 'counter.min_reset',
              metricType: MetricsDataType.Sum,
              valueExpression: 'Value',
            },
          ],
          from: metricSource.from,
          where: '',
          metricTables: {
            sum: DEFAULT_METRICS_TABLE.SUM,
            gauge: DEFAULT_METRICS_TABLE.GAUGE,
            histogram: DEFAULT_METRICS_TABLE.HISTOGRAM,
          },
          dateRange: [new Date(now), new Date(now + ms('20m'))],
          granularity: '10 minute',
          timestampValueExpression: metricSource.timestampValueExpression,
          connection: connection.id,
        },
        metadata,
      );

      const maxQuery = await renderChartConfig(
        {
          select: [
            {
              aggFn: 'max',
              metricName: 'counter.min_reset',
              metricType: MetricsDataType.Sum,
              valueExpression: 'Value',
            },
          ],
          from: metricSource.from,
          where: '',
          metricTables: {
            sum: DEFAULT_METRICS_TABLE.SUM,
            gauge: DEFAULT_METRICS_TABLE.GAUGE,
            histogram: DEFAULT_METRICS_TABLE.HISTOGRAM,
          },
          dateRange: [new Date(now), new Date(now + ms('20m'))],
          granularity: '10 minute',
          timestampValueExpression: metricSource.timestampValueExpression,
          connection: connection.id,
        },
        metadata,
      );
    });
  });
});
