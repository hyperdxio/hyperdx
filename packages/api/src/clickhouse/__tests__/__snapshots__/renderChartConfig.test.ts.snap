// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderChartConfig Query Events - Logs simple select + group by query logs 1`] = `
Array [
  Object {
    "ServiceName": "app",
    "count": "1",
  },
  Object {
    "ServiceName": "api",
    "count": "1",
  },
]
`;

exports[`renderChartConfig Query Events - Logs simple select + where query logs 1`] = `
Array [
  Object {
    "Body": "Oh no! Something went wrong!",
  },
]
`;

exports[`renderChartConfig Query Metrics - Gauge single avg gauge with group-by 1`] = `
Array [
  Object {
    "__hdx_time_bucket": "2022-01-05T00:00:00Z",
    "arrayElement(ResourceAttributes, 'host')": "host2",
    "avg(toFloat64OrDefault(toString(LastValue)))": 4,
  },
  Object {
    "__hdx_time_bucket": "2022-01-05T00:00:00Z",
    "arrayElement(ResourceAttributes, 'host')": "host1",
    "avg(toFloat64OrDefault(toString(LastValue)))": 6.25,
  },
  Object {
    "__hdx_time_bucket": "2022-01-05T00:05:00Z",
    "arrayElement(ResourceAttributes, 'host')": "host2",
    "avg(toFloat64OrDefault(toString(LastValue)))": 4,
  },
  Object {
    "__hdx_time_bucket": "2022-01-05T00:05:00Z",
    "arrayElement(ResourceAttributes, 'host')": "host1",
    "avg(toFloat64OrDefault(toString(LastValue)))": 80,
  },
]
`;

exports[`renderChartConfig Query Metrics - Gauge single avg gauge with where 1`] = `
Array [
  Object {
    "__hdx_time_bucket": "2022-01-05T00:00:00Z",
    "avg(toFloat64OrDefault(toString(LastValue)))": 6.25,
  },
  Object {
    "__hdx_time_bucket": "2022-01-05T00:05:00Z",
    "avg(toFloat64OrDefault(toString(LastValue)))": 80,
  },
]
`;

exports[`renderChartConfig Query Metrics - Gauge single max/avg/sum gauge 1`] = `
Array [
  Object {
    "__hdx_time_bucket": "2022-01-05T00:00:00Z",
    "avg(toFloat64OrDefault(toString(LastValue)))": 5.125,
  },
  Object {
    "__hdx_time_bucket": "2022-01-05T00:05:00Z",
    "avg(toFloat64OrDefault(toString(LastValue)))": 42,
  },
]
`;

exports[`renderChartConfig Query Metrics - Gauge single max/avg/sum gauge 2`] = `
Array [
  Object {
    "__hdx_time_bucket": "2022-01-05T00:00:00Z",
    "max(toFloat64OrDefault(toString(LastValue)))": 6.25,
  },
  Object {
    "__hdx_time_bucket": "2022-01-05T00:05:00Z",
    "max(toFloat64OrDefault(toString(LastValue)))": 80,
  },
]
`;

exports[`renderChartConfig Query Metrics - Gauge single max/avg/sum gauge 3`] = `
Array [
  Object {
    "__hdx_time_bucket": "2022-01-05T00:00:00Z",
    "sum(toFloat64OrDefault(toString(LastValue)))": 10.25,
  },
  Object {
    "__hdx_time_bucket": "2022-01-05T00:05:00Z",
    "sum(toFloat64OrDefault(toString(LastValue)))": 84,
  },
]
`;

exports[`renderChartConfig Query Metrics - Histogram should bucket correctly when grouping by a single attribute 1`] = `
Array [
  Object {
    "Value": 3.5714285714285716,
    "__hdx_time_bucket": "2022-01-05T00:00:00Z",
    "group": Array [
      "host-b",
    ],
  },
  Object {
    "Value": 8.382352941176471,
    "__hdx_time_bucket": "2022-01-05T00:00:00Z",
    "group": Array [
      "host-a",
    ],
  },
  Object {
    "Value": 3.5,
    "__hdx_time_bucket": "2022-01-05T00:02:00Z",
    "group": Array [
      "host-b",
    ],
  },
  Object {
    "Value": 4.95,
    "__hdx_time_bucket": "2022-01-05T00:02:00Z",
    "group": Array [
      "host-a",
    ],
  },
]
`;

exports[`renderChartConfig Query Metrics - Histogram should bucket correctly when grouping by multiple attributes 1`] = `
Array [
  Object {
    "Value": 2.916666666666667,
    "__hdx_time_bucket": "2022-01-05T00:00:00Z",
    "group": Array [
      "host-b",
      "service-2",
    ],
  },
  Object {
    "Value": 4.852941176470588,
    "__hdx_time_bucket": "2022-01-05T00:00:00Z",
    "group": Array [
      "host-a",
      "service-2",
    ],
  },
  Object {
    "Value": 8.75,
    "__hdx_time_bucket": "2022-01-05T00:00:00Z",
    "group": Array [
      "host-a",
      "service-1",
    ],
  },
  Object {
    "Value": 3.1578947368421053,
    "__hdx_time_bucket": "2022-01-05T00:00:00Z",
    "group": Array [
      "host-b",
      "service-1",
    ],
  },
  Object {
    "Value": 58.33333333333333,
    "__hdx_time_bucket": "2022-01-05T00:00:00Z",
    "group": Array [
      "host-a",
      "service-3",
    ],
  },
  Object {
    "Value": 6.25,
    "__hdx_time_bucket": "2022-01-05T00:00:00Z",
    "group": Array [
      "host-b",
      "service-3",
    ],
  },
  Object {
    "Value": 3.4090909090909087,
    "__hdx_time_bucket": "2022-01-05T00:02:00Z",
    "group": Array [
      "host-b",
      "service-1",
    ],
  },
  Object {
    "Value": 7.916666666666667,
    "__hdx_time_bucket": "2022-01-05T00:02:00Z",
    "group": Array [
      "host-a",
      "service-1",
    ],
  },
  Object {
    "Value": 3.25,
    "__hdx_time_bucket": "2022-01-05T00:02:00Z",
    "group": Array [
      "host-b",
      "service-3",
    ],
  },
  Object {
    "Value": 4.25,
    "__hdx_time_bucket": "2022-01-05T00:02:00Z",
    "group": Array [
      "host-a",
      "service-3",
    ],
  },
  Object {
    "Value": 4.75,
    "__hdx_time_bucket": "2022-01-05T00:02:00Z",
    "group": Array [
      "host-a",
      "service-2",
    ],
  },
  Object {
    "Value": 3.888888888888889,
    "__hdx_time_bucket": "2022-01-05T00:02:00Z",
    "group": Array [
      "host-b",
      "service-2",
    ],
  },
]
`;

exports[`renderChartConfig Query Metrics - Histogram should bucket correctly when no grouping is defined 1`] = `
Array [
  Object {
    "Value": 5.241935483870968,
    "__hdx_time_bucket": "2022-01-05T00:00:00Z",
  },
  Object {
    "Value": 4.40625,
    "__hdx_time_bucket": "2022-01-05T00:02:00Z",
  },
]
`;

exports[`renderChartConfig Query Metrics - Histogram two_timestamps_bounded histogram (p25) 1`] = `
Array [
  Object {
    "Value": 7.5,
    "__hdx_time_bucket": "2022-01-05T00:01:00Z",
  },
]
`;

exports[`renderChartConfig Query Metrics - Histogram two_timestamps_bounded histogram (p50) 1`] = `
Array [
  Object {
    "Value": 20,
    "__hdx_time_bucket": "2022-01-05T00:01:00Z",
  },
]
`;

exports[`renderChartConfig Query Metrics - Histogram two_timestamps_bounded histogram (p90) 1`] = `
Array [
  Object {
    "Value": 30,
    "__hdx_time_bucket": "2022-01-05T00:01:00Z",
  },
]
`;

exports[`renderChartConfig Query Metrics - Histogram two_timestamps_lower_bound_inf histogram (p50) 1`] = `
Array [
  Object {
    "Value": 0.5,
    "__hdx_time_bucket": "2022-01-05T00:01:00Z",
  },
]
`;

exports[`renderChartConfig Query Metrics - Sum calculates min_rate/max_rate correctly for sum metrics: maxSum 1`] = `
Array [
  Object {
    "Value": 24,
    "__hdx_time_bucket": "2022-01-05T00:00:00Z",
  },
  Object {
    "Value": 134,
    "__hdx_time_bucket": "2022-01-05T00:10:00Z",
  },
]
`;

exports[`renderChartConfig Query Metrics - Sum calculates min_rate/max_rate correctly for sum metrics: minSum 1`] = `
Array [
  Object {
    "Value": 15,
    "__hdx_time_bucket": "2022-01-05T00:00:00Z",
  },
  Object {
    "Value": 52,
    "__hdx_time_bucket": "2022-01-05T00:10:00Z",
  },
]
`;

exports[`renderChartConfig Query Metrics - Sum handles counter resets correctly for sum metrics 1`] = `
Array [
  Object {
    "Value": 15,
    "__hdx_time_bucket": "2022-01-05T00:00:00Z",
  },
  Object {
    "Value": 52,
    "__hdx_time_bucket": "2022-01-05T00:10:00Z",
  },
]
`;

exports[`renderChartConfig Query Metrics - Sum single sum rate 1`] = `
Array [
  Object {
    "Value": 19,
    "__hdx_time_bucket": "2022-01-05T00:00:00Z",
  },
  Object {
    "Value": 79,
    "__hdx_time_bucket": "2022-01-05T00:05:00Z",
  },
  Object {
    "Value": 5813,
    "__hdx_time_bucket": "2022-01-05T00:10:00Z",
  },
  Object {
    "Value": 78754,
    "__hdx_time_bucket": "2022-01-05T00:15:00Z",
  },
]
`;

exports[`renderChartConfig Query Metrics - Sum sum values as without rate computation 1`] = `
Array [
  Object {
    "Value": 950400,
    "__hdx_time_bucket": "2022-01-05T00:00:00Z",
  },
  Object {
    "Value": 1641600,
    "__hdx_time_bucket": "2022-01-05T00:10:00Z",
  },
]
`;

exports[`renderChartConfig aggFn numeric agg functions should handle numeric values as strings 1`] = `
Array [
  Object {
    "AVG(toFloat64OrDefault(toString(strVal)))": 0.5,
    "MAX(toFloat64OrDefault(toString(strVal)))": 3,
    "MIN(toFloat64OrDefault(toString(strVal)))": -1.1,
    "SUM(toFloat64OrDefault(toString(strVal)))": 2,
    "quantile(0.5)(toFloat64OrDefault(toString(strVal)))": 0.050000000000000044,
  },
]
`;

exports[`renderChartConfig aggFn numeric agg functions should use default values for other types 1`] = `
Array [
  Object {
    "AVG(toFloat64OrDefault(toString(strVal)))": 0,
    "MAX(toFloat64OrDefault(toString(strVal)))": 0,
    "MIN(toFloat64OrDefault(toString(strVal)))": 0,
    "SUM(toFloat64OrDefault(toString(strVal)))": 0,
    "quantile(0.5)(toFloat64OrDefault(toString(strVal)))": 0,
  },
]
`;
